sw/device/silicon_creator/lib/boot_data_functest.c

```diff
@@ -232,6 +256,66 @@ rom_error_t read_full_page_1_test(void) {
   return kErrorOk;
 }
 
+rom_error_t write_empty_test(void) {
+  erase_boot_data_pages();
+  RETURN_IF_ERROR(boot_data_write(&kTestBootData));
+  boot_data_t boot_data;
+  RETURN_IF_ERROR(boot_data_read(kLcStateProd, &boot_data));
+  RETURN_IF_ERROR(compare_boot_data(&kTestBootData, &boot_data));
+  return kErrorOk;
+}
+
+rom_error_t write_page_switch_test(void) {
+  erase_boot_data_pages();
+  boot_data_t boot_data_act;
+  boot_data_t boot_data_exp;
+  // Counter value starts from `kBootDataDefault.counter` defined in
+  // `boot_data.c` and is incremented before each write.
+  uint32_t counter_exp = 5;
+
+  // Write `kBootDataEntriesPerPage` + 1 entries to test the switch from page 0
+  // to page 1.
+  for (size_t i = 0; i < kBootDataEntriesPerPage + 1; ++i) {
+    RETURN_IF_ERROR(boot_data_write(&kTestBootData));
+    // Check `identifier`, `digest`, and `counter` fields.
+    RETURN_IF_ERROR(boot_data_read(kLcStateProd, &boot_data_act));
+    RETURN_IF_ERROR(check_boot_data(&boot_data_act, ++counter_exp));
+    if (i > 0) {
+      // Previous entry must be invalidated.
+      boot_data_t prev_entry;
+      read_boot_data(kBootDataPage0Base, i - 1, &prev_entry);
+      if (prev_entry.identifier != kBootDataInvalidatedIdentifier) {
+        return kErrorUnknown;
+      }
+    }
+  }
+  // Last written entry must be at entry 0 in page 1.
+  read_boot_data(kBootDataPage1Base, 0, &boot_data_exp);
+  CHECK(memcmp(&boot_data_act, &boot_data_exp, sizeof(boot_data_t)) == 0,
```

Comment:
nit: Since you are using `CHECK`, it may be work also using `CHECK_BUFFER`.

Reply:
Do you mind if we do this when migrating to the silicon_creator flash_ctrl driver to minimize toil? `CHECK_BUFFER` was recently updated in #9003 so it will need a rebase but I have a few branches on top of this change.

Reply:
Actually, I think this should have been "check and return `kErrorUnknown`". Removed `CHECK()` here.